service/tx/service.go

```diff
@@ -182,6 +187,171 @@ func (svc *Service) CreateRawSendPool(args client.SendPoolTxRequest, reply *clie
+
+	allegation := penalization.AllegationVote{
+		Address:   validator.Address,
+		RequestID: args.RequestID,
+		Choice:    args.Choice,
+	}
+
+	data, err := allegation.Marshal()
+	if err != nil {
+		svc.logger.Error("error in serializing vote object", err)
+		return codes.ErrSerialization
+	}
+
+	uuidNew, _ := uuid.NewUUID()
+	feeAmount := svc.feeOpt.MinFee()
+
+	tx := action.RawTx{
+		Type: action.ALLEGATION_VOTE,
+		Data: data,
+		Fee: action.Fee{
+			Price: action.Amount{
+				Currency: "OLT",
+				Value:    *feeAmount.Amount,
+			},
+			Gas: 100000,
+		},
+		Memo: uuidNew.String(),
+	}
+
+	packet, err := serialize.GetSerializer(serialize.NETWORK).Serialize(tx)
+	if err != nil {
+		svc.logger.Error("error in serializing send transaction", err)
+		return codes.ErrSerialization
+	}
+
+	*reply = client.VoteReply{RawTx: packet}
+
+	return nil
+}
+
+func (svc *Service) Allegation(args client.AllegationRequest, reply *client.AllegationReply) error {
+	mv, err := svc.validators.Get(args.MaliciousAddress)
+	if err != nil {
+		svc.logger.Errorf("malicious validator for address %s not found\n", args.MaliciousAddress)
+		return codes.ErrBadOwner
+	}
+
+	if svc.evidenceStore.IsFrozenValidator(mv.Address) {
+		svc.logger.Error("validator already frozen", mv.Address)
+		return codes.ErrFrozenValidator
+	}
+
+	validator, err := svc.validators.Get(args.Address)
+	if err != nil {
+		svc.logger.Errorf("validator for address %s not found\n", args.Address)
+		return codes.ErrBadOwner
+	}
+
+	if svc.evidenceStore.IsFrozenValidator(validator.Address) {
+		svc.logger.Error("validator could no perform allegation while frosted", validator.Address)
+		return codes.ErrFrozenValidator
+	}
+
+	allegation := penalization.Allegation{
+		ValidatorAddress: validator.Address,
+		MaliciousAddress: mv.Address,
+		BlockHeight:      args.BlockHeight,
+		ProofMsg:         args.ProofMsg,
+	}
+
+	data, err := allegation.Marshal()
+	if err != nil {
+		svc.logger.Error("error in serializing allegation object", err)
+		return codes.ErrSerialization
+	}
+
+	uuidNew, _ := uuid.NewUUID()
+	feeAmount := svc.feeOpt.MinFee()
+
+	tx := action.RawTx{
+		Type: action.ALLEGATION,
+		Data: data,
+		Fee: action.Fee{
+			Price: action.Amount{
+				Currency: "OLT",
+				Value:    *feeAmount.Amount,
+			},
+			Gas: 100000,
+		},
+		Memo: uuidNew.String(),
+	}
+
+	packet, err := serialize.GetSerializer(serialize.NETWORK).Serialize(tx)
+	if err != nil {
+		svc.logger.Error("error in serializing send transaction", err)
+		return codes.ErrSerialization
+	}
+
+	*reply = client.AllegationReply{RawTx: packet}
+
+	return nil
+}
+
+func (svc *Service) Release(args client.ReleaseRequest, reply *client.ReleaseReply) error {
+	validator, err := svc.validators.Get(args.Address)
+	if err != nil {
+		svc.logger.Errorf("validator for address %s not found\n", args.Address)
+		return codes.ErrBadOwner
+	}
+
+	if !svc.evidenceStore.IsFrozenValidator(validator.Address) {
```

Comment:
Have you checked this calling Tx being executed under the same the validator node denoted by args.Address? You can leave the client.ReleaseRequest struct empty and let this Release function to get the node address and fill in penalization.Release{} struct.